staging/src/k8s.io/client-go/tools/events/event_broadcaster.go

```diff
@@ -0,0 +1,294 @@
+		if eventSerie != nil {
+
+			if eventSerie.LastObservedTime.Time.Add(6 * time.Minute).Before(time.Now()) {
+
+				if recordEvent(eventBroadcaster.sink, event, true) {
+					delete(eventBroadcaster.eventCache, isomorphicKey)
+				}
+			}
+		}
+
+	}
+}
+
+// NewRecorder returns an EventRecorder that records events with the given event source.
+func (eventBroadcaster *eventBroadcasterImpl) NewRecorder(scheme *runtime.Scheme, reportingController string) EventRecorder {
+
+	hostname, _ := os.Hostname()
+	reportingInstance := reportingController + "-" + hostname
+
+	return &recorderImpl{scheme, reportingController, reportingInstance, eventBroadcaster.Broadcaster, clock.RealClock{}}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) recordToSink(event *v1beta1.Event, clock clock.Clock) {
+	// Make a copy before modification, because there could be multiple listeners.
+	// Events are safe to copy like this.
+	eventCopy := *event
+	event = &eventCopy
+
+	go func() {
+		eventKey := getKey(event)
+		eventBroadcaster.mu.RLock()
+		isomorphicEvent, isIsomorphic := eventBroadcaster.eventCache[eventKey]
+		eventBroadcaster.mu.RUnlock()
+
+		if isIsomorphic {
+
+			if isomorphicEvent.Series != nil {
+				isomorphicEvent.Series.Count++
+				isomorphicEvent.EventTime = metav1.MicroTime{Time: clock.Now()}
+			} else {
+				isomorphicEvent.Series = &v1beta1.EventSeries{
+					Count:            1,
+					LastObservedTime: metav1.MicroTime{Time: clock.Now()},
+				}
+				eventBroadcaster.attemptRecording(isomorphicEvent)
+			}
+
+		} else {
+			eventBroadcaster.mu.Lock()
+			eventBroadcaster.eventCache[eventKey] = event
+			eventBroadcaster.mu.Unlock()
+			eventBroadcaster.attemptRecording(event)
+		}
+
+	}()
+
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) attemptRecording(event *v1beta1.Event) {
+	tries := 0
+
+	for {
+		if recordEvent(eventBroadcaster.sink, event, event.Series != nil) {
+			break
+		}
+		tries++
+		if tries >= maxTriesPerEvent {
+			klog.Errorf("Unable to write event '%#v' (retry limit exceeded!)", event)
+			break
+		}
+		// Randomize sleep so that various clients won't all be
+		// synced up if the master goes down.
+		time.Sleep(time.Duration(float64(eventBroadcaster.sleepDuration) * eventBroadcaster.randGen.Float64()))
+	}
+
+}
+
+func recordEvent(sink EventSink, event *v1beta1.Event, isEventSerie bool) bool {
+	var newEvent *v1beta1.Event
+	var err error
+	if isEventSerie {
+
+		patch, err := createPatchBytesForSeries(event)
+		if err != nil {
+			klog.Errorf("Unable to calculate diff, no merge is possible: %v", err)
+			return true
+		}
+		newEvent, err = sink.Patch(event, patch)
+	}
+	// Update can fail because the event may have been removed and it no longer exists.
+	if !isEventSerie || (isEventSerie && util.IsKeyNotFoundError(err)) {
+		// Making sure that ResourceVersion is empty on creation
+		event.ResourceVersion = ""
+		newEvent, err = sink.Create(event)
+	}
+	if err == nil {
+
+		event = newEvent
```

Comment:
I'm not a fan of this pattern. If you want to return the recorded version of the event, you should do that explicitly in the function signature.